File: SolutionForCritique/Python/patient.py

class Patient(object):
    
    def __init__(self, medicines = None):
        self._medicines = medicines or []
    
    def add_medicine(self, medicine):
        self._medicines.append(medicine)
    
    def clash(self, medicine_names, days_back):
        medicines = [medicine for medicine in self._medicines if medicine.name in medicine_names]
        if not medicines:
            return set()
        dates = set.intersection(*[set(medicine.dates_prescribed_in_effective_range(days_back)) for medicine in medicines])
        return dates

File: C++/Patient.cpp
#include "Patient.hpp"

#include <chrono>
#include <string>
#include <vector>

using namespace std;
using std::chrono::system_clock;

Patient::Patient(string const & name) : name_(name)
{
}

vector<time_t> Patient::clash(vector<string> medicineNames, int daysBack) const
{
    /* TODO: write this code */
    vector<time_t> dates;
    return dates;
}

void Patient::addMedicine(Medicine medicine)
{
    medicines_.push_back(medicine);
}
